Docs: add topic for lifecycle of a plugin (#457)
grafana/plugin-tools: docusaurus/docs/introduction/plugin-lifecycle.md

---
id: plugin-lifecycle
title: Life cycle of a plugin
description: Learn about the life cycle of a Grafana plugin.
keywords:
  - grafana
  - plugins
  - plugin
  - lifecycle
  - life cycle
  - loading
  - unloading
  - installation
sidebar_position: 2.5
---

# Life cycle of a plugin

> This document describes the various phases of a plugin such as installation and loading. We will describe the differences in a plugin's life cycle depending on its type and whether or not it has a [backend](backend.md).  

Reviewer: <img width="1273" alt="Screenshot 2023-10-26 at 20 16 01" src="https://github.com/grafana/plugin-tools/assets/19860021/b0cde578-964f-4fb0-8b9a-0db90ce3477d">

Seeing this a lot on the backend link... maybe fine, would advise pushing the PR to dev using https://github.com/grafana/plugin-tools/actions/workflows/deploy-to-developer-portal-dev.yml to verify before merging.

Author: compilation working for me now, after jackw's PR merged.

:::note

> The life cycle of plugins is tracked in-memory and is not persisted in Grafana’s database. 

Reviewer: I think it would be good to explain why this matters to the reader. I _believe_ it means that it happens very time the grafana server is restarted. So something like "This means that the phases described below occur every time the server is restarted." Someone on the team could confirm/amend this.

If this indeed the case, it might be better to move this note to the Loading plugins section.

Author: done. @marefr - can you add to/improve on david's suggestion?

:::

## Installing and uninstalling a plugin

For instructions to install or uninstall plugins, see our documentation at [Plugin administration](https://grafana.com/docs/grafana/latest/administration/plugin-management/#install-grafana-plugins).

Upon installation, a plugin is extracted to the _plugin directory_ on the filesystem. Similarly, uninstalling a plugin deletes the files from the same directory.  

:::note

Use of the Grafana CLI to install or uninstall a plugin requires you to restart Grafana for the change to take effect. To avoid restarting Grafana, you can install a plugin from within the Grafana [plugin catalog](https://grafana.com/plugins/) during runtime.

:::

## Loading plugins

Understanding the different phases involved when Grafana is starting a plugin may help you better understand plugin usage and to debug any unexpected behavior during loading.

> 
> All plugins are initialized either when Grafana starts up or when a plugin has been installed/uninstalled during runtime.

Reviewer 2: With earlier suggestion moving this to the beginning

```suggestion
```

> For [backend](./backend.md) plugins, there is an additional initialization process (see [Phase 4](#phase-4-backend-plugin-initialization)).

Reviewer: this link doesnt work as its now phase 3

- Angular detection: Given that [Angular is deprecated](https://grafana.com/docs/grafana/latest/developers/angular_deprecation/), if Angular support is disabled and Angular is detected in the plugin, then we record an error and don't allow the plugin to be loaded.

### Phase 3. Backend plugin initialization

For any verified plugin that has a backend, Grafana configures the backend client to use HashiCorp’s Go Plugin System over RPC.

### Phase 4. Registration

All verified plugins are registered in an in-memory registry. From now on, the plugin is available within Grafana and so are referred to as _registered plugins_.

Registered plugins show as installed in the catalog and appear in views for selecting a panel or data source in a dashboard.

### Phase 5. Start backend plugin

For registered plugins that have a backend, Grafana starts to run the backend binary as a separate process using HashiCorp’s Go Plugin System over RPC. The supported plugin protocol and version is negotiated between Grafana (client) and the plugin (server) to give Grafana an understanding of the plugin's capabilities. 

> A Grafana backend plugin has its own separate life cycle. So long as the backend plugin is running, Grafana will make sure to restart the backend plugin in case it crashes or is terminated. When Grafana is shut down, the backend processes is then terminated.  

Reviewer: > A Grafana backend plugin has its own separate life cycle

I'm not sure this statement is clear enough, separate to what?

> So long as the backend plugin is running, Grafana will make sure to restart the backend plugin in case it crashes or is terminated.

I find this a little unclear, if the backend plugin crashes, it is no longer running and needs to be restarted, so the opening statement feels confusing.

Author: @marefr - again, could use your expertise to clarify these points

### Phase 6. Client-side loading 

> After Grafana has started and the HTTP API is running, Grafana users receive the server-side rendered index page containing so-called bootstrap data. This data includes the list of available plugins and a URI to a `module.js` file that Grafana uses to instantiate the plugin. 

Reviewer: Nit: we could link to https://grafana.com/docs/grafana/latest/developers/http_api/

Author: done.

---
id: plugin-lifecycle
title: Life cycle of a plugin
description: Learn about the life cycle of a Grafana plugin.
keywords:
  - grafana
  - plugins
  - plugin
  - lifecycle
  - life cycle
  - loading
  - unloading
  - installation

> sidebar_position: 2.5

Reviewer: We don't have to fix this now, but suggest we multiply the sidebar positions by 10 to make simpler as in https://github.com/grafana/plugin-tools/tree/main/docusaurus/docs/tutorials

Author: yeah, I will make this change in a batch at a later date. for now, fractions.

Author: when i do that batch, that would also be a good time to fine-tune any sidebar ordering for which we want to improve the flow of topics.

---

# Life cycle of a plugin

> This document describes the various phases of a plugin including installation and loading. We will describe the differences in a plugin's life cycle depending on its type and whether or not it has a backend.  

Reviewer: Could link to the backend topic here

Author: done.

:::note

The life cycle of plugins is tracked in-memory and is not persisted in Grafana’s database. 

:::

## Installing and uninstalling a plugin

For instructions to install or uninstall plugins, see our documentation at [Plugin administration](https://grafana.com/docs/grafana/latest/administration/plugin-management/#install-grafana-plugins).

Upon installation, a plugin is extracted to the _plugin directory_ on the filesystem. Similarly, uninstalling a plugin deletes the files from the same directory.  

:::note

Use of the Grafana CLI to install or uninstall a plugin requires you to restart Grafana for the change to take effect. To avoid restarting Grafana, you can install a plugin from within the Grafana [plugin catalog](https://grafana.com/plugins/) during runtime.

:::

## Loading plugins

> Understanding the different phases involved when Grafana is starting a plugin may help you better understand plugin usage and to debug any unexpected behavior during loading.

Reviewer: Is there anything more we could offer for users to `debug any unexpected behavior during loading`

What are the errors they may see, where will they see them?

Author: @marefr - Any thoughts on this? I would like to see more detail, or perhaps we remove?

Reviewer 2: I started to look at what information is logged when Grafana starts up and noticed that we lacked some standard info log messages why I opened https://github.com/grafana/grafana/pull/76896. 

I think it make sense to keep this, but I would suggest referring to inspecting the Grafana server log somehow with optional debug level enabled. 

> What are the errors they may see, where will they see them?

I think it might be hard to explain in detail any errors that could happen.

Something like below perhaps. WDYT?

```suggestion
Plugins are loaded either when Grafana starts up or when a plugin has been installed/uninstalled during runtime. 

Understanding the different phases involved when Grafana is loading a plugin may help you better understand plugin usage and troubleshoot any unexpected behavior, such as why a certain plugin is not marked as installed in the plugins catalog and/or for use within Grafana even though you've installed it. You can check the [Grafana server log](https://grafana.com/docs/grafana/latest/troubleshooting/#troubleshoot-with-logs) for any unexpected errors or details related to loading a plugin. In addition, you can enable even more details by changing the [log level to debug](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#log).
```

> ### Phase 1. Plugin initialization

Reviewer: It's not clear what has happened as a result of the plugin being initialised?

Reviewer 2: As described below, Grafana starts up or a plugin has been installed/uninstalled during runtime. 

Maybe skip this _Phase 1_ and explain in _Loading plugins_ these are the phases involved when Grafana starts up or when a plugin has been installed/uninstalled during runtime?

Author: Phase 1 skipped for reason noted above.

All plugins are initialized either when Grafana starts up or when a plugin has been installed/uninstalled during runtime.

For [backend](./backend.md) plugins, there is an additional initialization process (see [Phase 4](#phase-4-backend-plugin-initialization)).

### Phase 2. Plugin discovery 

Grafana starts to discover which plugins are installed by scanning directories on the file system for every `plugin.json`.

### Phase 3. Plugin loading 

> All plugins that were discovered in the discovery phase are checked to make sure they’re valid and that they have a valid signature. Valid plugins are referred to as _verified plugins_. 

Reviewer: ```suggestion
All plugins that were discovered in the discovery phase are checked to make sure they’re valid and that they have a valid [signature](https://grafana.com/docs/grafana/latest/administration/plugin-management/#plugin-signatures). Valid plugins are referred to as _verified plugins_. 
```

Reviewer: Beyond a signature check, what else is validated?

Reviewer 2: Not sure we need to mention that in detail. But to mention what we currently have
- Logs a warning message if a module.js file is missing
- Angular detection: If angular support is disabled and angular detected in plugin we record an error and doesn't allow this plugin to be loaded

Author: I didnn't add module.js error, but I added Angular deprecation because I think that's something that we want to publicize whenever it makes sense.

### Phase 4. Backend plugin initialization

> For any verified plugin that has a backend, Grafana configures a backend client. Depending on the plugin's classification, one of these things will happen: 

Reviewer: We've not defined Core plugins. Either needs to be here or in the introduction

Reviewer 2: To simplify I suggest skip core plugins specific details. WDYT?

> For any verified plugin that has a backend, Grafana configures the backend client to use HashiCorp’s Go Plugin System over RPC.

Author: @sympatheticmoose - Good catch, but there is a third option, which is not to mention classifications at all.
@marefr - I agree because as we discussed an explanation of plugin classifications is primarily a distinction important to internal audience not needed for public document.

- **Core plugins** - the backend client is already configured and nothing more needs to be done.
- **External plugins** - Grafana configures the backend client to use HashiCorp’s Go Plugin System over RPC.

### Phase 5. Registration

> All verified plugins are registered in an in-memory registry. From now on, the plugin is available within Grafana and so are referred to as _registered plugins_.

Reviewer: What does it mean to be available within Grafana? Is it now that the plugin would show as installed in the catalog and appear in views for selecting a panel or data source in a dashboard

Reviewer 2: Yes exactly

Author: Good additional info.

> ### Phase 6. Start backend plugin (backend plugins only)

Reviewer 2: I think Start backend plugin on its own describes this is backend plugins only 

```suggestion
### Phase 6. Start backend plugin
```

Author: fixed.

> For registered external plugins that have a backend, Grafana starts to run the backend binary as a separate process using HashiCorp’s Go Plugin System over RPC. The supported plugin protocol and version is negotiated between Grafana (client) and the plugin (server) to give Grafana an understanding of the plugin's capabilities.  

Reviewer 2: ```suggestion
For registered plugins that have a backend, Grafana starts to run the backend binary as a separate process using HashiCorp’s Go Plugin System over RPC. The supported plugin protocol and version is negotiated between Grafana (client) and the plugin (server) to give Grafana an understanding of the plugin's capabilities.  
```

Author: fixed.

> A Grafana backend plugin has its own separate life cycle. So long as the backend plugin is running, Grafana will make sure to restart the backend plugin in case it crashes or is killed. When Grafana is shut down, the backend processes is then terminated.  

Reviewer 2: Less hard phrasing perhaps :) 

```suggestion
A Grafana backend plugin has its own separate life cycle. So long as the backend plugin is running, Grafana will make sure to restart the backend plugin in case it crashes or is terminated. When Grafana is shut down, the backend processes is then terminated.  
```

Author: good catch.

### Phase 7. Client-side loading 

After Grafana has started and the HTTP API is running, Grafana users receive the server-side rendered index page containing so-called bootstrap data. This data includes the list of available plugins and a URI to a `module.js` file that Grafana uses to instantiate the plugin. 

> When the user interacts with a UI that requires a plugin, Grafana will _lazy load_ the plugin's `module.js` file: 

Reviewer: Panels are not discussed here, should they be?

Reviewer 2: Probably should be mentioned

Author: added. please re-review text.

> - **Data-source plugins** - A data-source plugin could be loaded in more than one way. For instance, it could be loaded in the Explore page if the user selects the data source in the dropdown, or if they load a dashboard containing a plugin data source.

Reviewer 2: Should it be Datasource plugins or Data source plugins?

Author: I know this looks weird, but data-source is our current Grafana style for adjectives. See https://grafana.com/docs/writers-toolkit/write/style-guide/word-list/#d

- **App plugins** - Apps have two different loading modes: _lazy_ and _pre-load_. Lazy app plugins load only when the user accesses the App menu item directly. Pre-load app plugins load with the Grafana app and can execute code as soon as the page loads.

:::note

Each plugin is loaded only once but its objects are initialized multiple times. For example, a dashboard with 10 different panel plugins will load 10 plugins with an instance of each. A dashboard with 10 panels of the same plugin will load the plugin once and have 10 instances.

:::

> ## Plugin deprecation

Reviewer: Deprecation feels separate to the rest of this as it is more about its availability within the Plugins Catalog. I would remove

Author: removed.

:::

### Phase 5. Registration

All verified plugins are registered in an in-memory registry. From now on, the plugin is available within Grafana and so are referred to as _registered plugins_.

### Phase 6. Start backend plugin (backend plugins only)

For registered external plugins that have a backend, Grafana starts to run the backend binary as a separate process using HashiCorp’s Go Plugin System over RPC. The supported plugin protocol and version is negotiated between Grafana (client) and the plugin (server) to give Grafana an understanding of the plugin's capabilities.  

A Grafana backend plugin has its own separate life cycle. So long as the backend plugin is running, Grafana will make sure to restart the backend plugin in case it crashes or is killed. When Grafana is shut down, the backend processes is then terminated.  

### Phase 7. Client-side loading 

After Grafana has started and the HTTP API is running, Grafana users receive the server-side rendered index page containing so-called bootstrap data. This data includes the list of available plugins and a URI to a `module.js` file that Grafana uses to instantiate the plugin. 

> When the user interacts with a UI that requires a plugin, Grafana _lazy-loads_ the plugin’s fetch request: 

Reviewer 3: ```suggestion
When the user interacts with a UI that requires a plugin, Grafana will _lazy load_ the plugin's `module.js` file: 
```

### Phase 1. Plugin initialization

All plugins are initialized either when Grafana starts up or when a plugin has been installed/uninstalled during runtime.

For [backend](./backend.md) plugins, there is an additional initialization process (see [Phase 4](#phase-4-backend-plugin-initialization)).

### Phase 2. Plugin discovery 

Grafana starts to discover which plugins are installed by scanning directories on the file system for every `plugin.json`.

### Phase 3. Plugin loading 

All plugins that were discovered in the discovery phase are checked to make sure they’re valid and that they have a valid signature. Valid plugins are referred to as _verified plugins_. 

### Phase 4. Backend plugin initialization

For any verified plugin that has a backend, Grafana configures a backend client. Depending on the plugin's classification, one of these things will happen: 
- **Core plugins** - the backend client is already configured and nothing more needs to be done.
- **External plugins** - Grafana configures the backend client to use HashiCorp’s Go Plugin System over RPC.

> 
> :::note
> 
> If a remote plugin has been configured to connect to a remote service, then the backend client will be configured to connect to a remote gRPC service instead.

Reviewer 2: ```suggestion
```
